node/cn/peer_set.go

```diff
@@ -0,0 +1,435 @@
+}
+
+func (ps *peerSet) ENWithoutBlock(hash common.Hash) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.enpeers))
+	for _, p := range ps.enpeers {
+		if !p.KnowsBlock(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+func (ps *peerSet) TypePeers(nodetype p2p.ConnType) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+	list := make([]Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		if p.ConnType() == nodetype {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+// PeersWithoutTx retrieves a list of peers that do not have a given transaction
+// in their set of known hashes.
+func (ps *peerSet) PeersWithoutTx(hash common.Hash) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		if !p.KnowsTx(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+func (ps *peerSet) TypePeersWithoutTx(hash common.Hash, nodetype p2p.ConnType) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		if p.ConnType() == nodetype && !p.KnowsTx(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+func (ps *peerSet) TypePeersWithTx(hash common.Hash, nodetype p2p.ConnType) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		if p.ConnType() == nodetype && p.KnowsTx(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+func (ps *peerSet) AnotherTypePeersWithoutTx(hash common.Hash, nodetype p2p.ConnType) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		if p.ConnType() != nodetype && !p.KnowsTx(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+// TODO-Klaytn drop or missing tx
+func (ps *peerSet) AnotherTypePeersWithTx(hash common.Hash, nodetype p2p.ConnType) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		if p.ConnType() != nodetype && p.KnowsTx(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+func (ps *peerSet) CNWithoutTx(hash common.Hash) []Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]Peer, 0, len(ps.cnpeers))
+	for _, p := range ps.cnpeers {
+		if !p.KnowsTx(hash) {
+			list = append(list, p)
+		}
+	}
+	return list
+}
+
+// BestPeer retrieves the known peer with the currently highest total blockscore.
+func (ps *peerSet) BestPeer() Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	var (
+		bestPeer Peer
+		bestTd   *big.Int
+	)
+	for _, p := range ps.peers {
+		if _, td := p.Head(); bestPeer == nil || td.Cmp(bestTd) > 0 {
+			bestPeer, bestTd = p, td
+		}
+	}
+	return bestPeer
+}
+
+// RegisterValidator registers a validator.
+func (ps *peerSet) RegisterValidator(connType p2p.ConnType, validator p2p.PeerTypeValidator) {
```

Comment:
Can I know a reason to introduce `RegisterValidator` to the peerSet?

Reply:
@jeonghonah Sure. Before the change, `ProtocolManager` can directly access to `ChannelManager` of `peerSet`, however, as `peerSet` has been changed into `PeerSet` interface, we need an additional interface, `RegisterValidator`.